feat(map): use targeting like syntax for configuration (#191)
saltstack-formulas/openssh-formula: docs/map.jinja.rst

2. parameters of the formulas in the file backend `roots`_

Configure the fileserver backends
`````````````````````````````````

I configure the `fileserver`_ backends to serve:

1. files from `roots`_ first
2. `gitfs`_ repositories last

Create the file ``/etc/salt/master.d/fileserver.conf`` and restart the ``master``:

.. code-block:: yaml

    ---
    ##
    ## file server
    ##
    fileserver_backend:
      # parameters values and override
      - roots
      # formulas
      - gitfs

    # The files in this directory will take precedence over git repositories
    file_roots:
      base:
        - /srv/salt

    # List of formulas I'm using
    gitfs_remotes:
      - https://github.com/saltstack-formulas/template-formula.git:
        - base: v4.1.1
      - https://github.com/saltstack-formulas/openssh-formula.git:
        - base: v2.0.1
    ...

Create per DNS configuration for ``TEMPLATE`` formula
`````````````````````````````````````````````````````

Now, we can provides the per DNS domain name configuration files for the ``TEMPLATE`` formulas under ``/srv/salt/TEMPLATE/parameters/``.

> We create the directory for ``dns:domain`` grain and we add a symlink for the ``domain`` grain which is extracted from the minion ``id``:

Reviewer: Something to bear in mind here is that you can't use a colon in filenames on Windows.

Author: Thanks a lot @dafyddj for the review.

I was not aware about the colon issue for `dns:domain` but it's only a problem if the master is on windows, right?

I'll make some tests tomorrow to be sure it's working correctly when the master is a GNU/Linux with a Windows minion. The YAML file is downloaded and should be in the minion cash but I don't know if the path naming could be an issue.

Reviewer: The Salt Master does run on Windows (apparently), but also this could easily affect masterless minions on Windows.

Author: Ok, we have an issue on Windows but I think it's more a salt bug:

```
2020-10-13 17:46:01,117 [salt.loaded.int.module.logmod:42  ][DEBUG   ][3024] map.jinja: load parameters from sources:
- Y:G@osarch
- Y:G@os_family
- Y:G@os
- Y:G@osfinger
- C@ad:lookup
- C@ad
- Y:G@dns:domain
- Y:G@id
2020-10-13 17:46:01,117 [salt.loaded.int.module.logmod:42  ][DEBUG   ][3024] map.jinja: load per formula default values from ad/parameters/defaults.yaml
2020-10-13 17:46:01,117 [salt.fileclient  :1152][DEBUG   ][3024] In saltenv 'base', looking at rel_path 'ad/parameters/defaults.yaml' to resolve 'salt://ad/parameters/defaults.yaml'
2020-10-13 17:46:01,117 [salt.fileclient  :1161][DEBUG   ][3024] In saltenv 'base', ** considering ** path 'c:\salt\var\cache\salt\minion\files\base\ad\parameters\defaults.yaml' to resolve 'salt://ad/parameters/defaults.yaml'
2020-10-13 17:46:01,148 [salt.utils.lazy  :102 ][DEBUG   ][3024] LazyLoaded config.get
2020-10-13 17:46:01,148 [salt.loaded.int.module.logmod:42  ][DEBUG   ][3024] map.jinja: lookup 'osarch' with 'grains.get'
2020-10-13 17:46:01,148 [salt.utils.lazy  :102 ][DEBUG   ][3024] LazyLoaded grains.get
2020-10-13 17:46:01,148 [salt.loaded.int.module.logmod:42  ][DEBUG   ][3024] map.jinja: load parameters from file ad/parameters//osarch.yaml
2020-10-13 17:46:01,148 [salt.fileclient  :1129][DEBUG   ][3024] Could not find file 'salt://ad/parameters//osarch.yaml' in saltenv 'base'
2020-10-13 17:46:01,148 [salt.loaded.int.module.logmod:42  ][DEBUG   ][3024] map.jinja: lookup 'os_family' with 'grains.get'
2020-10-13 17:46:01,163 [salt.loaded.int.module.logmod:42  ][DEBUG   ][3024] map.jinja: load parameters from file ad/parameters/os_family/Windows.yaml
2020-10-13 17:46:01,163 [salt.fileclient  :1129][DEBUG   ][3024] Could not find file 'salt://ad/parameters/os_family/Windows.yaml' in saltenv 'base'
2020-10-13 17:46:01,163 [salt.loaded.int.module.logmod:42  ][DEBUG   ][3024] map.jinja: lookup 'os' with 'grains.get'
2020-10-13 17:46:01,163 [salt.loaded.int.module.logmod:42  ][DEBUG   ][3024] map.jinja: load parameters from file ad/parameters/os/Windows.yaml
2020-10-13 17:46:01,163 [salt.fileclient  :1129][DEBUG   ][3024] Could not find file 'salt://ad/parameters/os/Windows.yaml' in saltenv 'base'
2020-10-13 17:46:01,163 [salt.loaded.int.module.logmod:42  ][DEBUG   ][3024] map.jinja: lookup 'osfinger' with 'grains.get'
2020-10-13 17:46:01,163 [salt.loaded.int.module.logmod:42  ][DEBUG   ][3024] map.jinja: load parameters from file ad/parameters/osfinger/Windows-10.yaml
2020-10-13 17:46:01,163 [salt.fileclient  :1129][DEBUG   ][3024] Could not find file 'salt://ad/parameters/osfinger/Windows-10.yaml' in saltenv 'base'
2020-10-13 17:46:01,163 [salt.loaded.int.module.logmod:42  ][DEBUG   ][3024] map.jinja: retrieve 'ad:lookup' with 'config.get', merge: strategy='None'
2020-10-13 17:46:01,163 [salt.loaded.int.module.logmod:42  ][DEBUG   ][3024] map.jinja: merge 'ad:lookup' retrieved with 'config.get', merge: strategy='smart', lists='False'
2020-10-13 17:46:01,163 [salt.utils.lazy  :102 ][DEBUG   ][3024] LazyLoaded slsutil.merge
2020-10-13 17:46:01,163 [salt.loaded.int.module.logmod:42  ][DEBUG   ][3024] map.jinja: retrieve 'ad' with 'config.get', merge: strategy='None'
2020-10-13 17:46:01,163 [salt.loaded.int.module.logmod:42  ][DEBUG   ][3024] map.jinja: merge 'ad' retrieved with 'config.get', merge: strategy='smart', lists='False'
2020-10-13 17:46:01,163 [salt.loaded.int.module.logmod:42  ][DEBUG   ][3024] map.jinja: lookup 'dns:domain' with 'grains.get'
2020-10-13 17:46:01,163 [salt.loaded.int.module.logmod:42  ][DEBUG   ][3024] map.jinja: load parameters from file ad/parameters/dns:domain/dompedago.etb1.lan.yaml
2020-10-13 17:46:01,179 [salt.fileclient  :1129][DEBUG   ][3024] Could not find file 'salt://ad/parameters/dns_domain/dompedago.etb1.lan.yaml' in saltenv 'base'
2020-10-13 17:46:01,179 [salt.loaded.int.module.logmod:42  ][DEBUG   ][3024] map.jinja: lookup 'id' with 'grains.get'
2020-10-13 17:46:01,179 [salt.loaded.int.module.logmod:42  ][DEBUG   ][3024] map.jinja: load parameters from file ad/parameters/id/PC-423853.dompedago.etb1.lan.yaml
2020-10-13 17:46:01,179 [salt.fileclient  :1129][DEBUG   ][3024] Could not find file 'salt://ad/parameters/id/PC-423853.dompedago.etb1.lan.yaml' in saltenv 'base'
2020-10-13 17:46:01,179 [salt.loaded.int.module.logmod:42  ][DEBUG   ][3024] map.jinja: save parameters in variable 'mapdata'
```

The minion mangle the `salt://` URL:

```
2020-10-13 17:46:01,163 [salt.loaded.int.module.logmod:42  ][DEBUG   ][3024] map.jinja: load parameters from file ad/parameters/dns:domain/dompedago.etb1.lan.yaml
2020-10-13 17:46:01,179 [salt.fileclient  :1129][DEBUG   ][3024] Could not find file 'salt://ad/parameters/dns_domain/dompedago.etb1.lan.yaml' in saltenv 'base'
```

When I create a link on the [fileserver](https://docs.saltstack.com/en/latest/ref/file_server):

```
ln -s /srv/salt/ad/parameters/dns:domain /srv/salt/ad/parameters/dns_domain
```

I got the following log:

```
2020-10-13 17:55:58,101 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4924] map.jinja: lookup 'dns:domain' with 'grains.get'
2020-10-13 17:55:58,101 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4924] map.jinja: load parameters from file ad/parameters/dns:domain/dompedago.etb1.lan.yaml
2020-10-13 17:55:58,117 [salt.fileclient  :1152][DEBUG   ][4924] In saltenv 'base', looking at rel_path 'ad/parameters/dns_domain/dompedago.etb1.lan.yaml' to resolve 'salt://ad/parameters/dns_domain/dompedago.etb1.lan.yaml'
2020-10-13 17:55:58,117 [salt.fileclient  :1161][DEBUG   ][4924] In saltenv 'base', ** considering ** path 'c:\salt\var\cache\salt\minion\files\base\ad\parameters\dns_domain\dompedago.etb1.lan.yaml' to resolve 'salt://ad/parameters/dns_domain/dompedago.etb1.lan.yaml'
2020-10-13 17:55:58,117 [salt.fileclient  :1179][DEBUG   ][4924] Fetching file from saltenv 'base', ** attempting ** 'salt://ad/parameters/dns_domain/dompedago.etb1.lan.yaml'
2020-10-13 17:55:58,117 [salt.fileclient  :1207][DEBUG   ][4924] No dest file found
2020-10-13 17:55:58,132 [salt.fileclient  :1295][INFO    ][4924] Fetching file from saltenv 'base', ** done ** 'ad/parameters/dns_domain/dompedago.etb1.lan.yaml'
2020-10-13 17:55:58,132 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4924] map.jinja: lookup 'id' with 'grains.get'
```

So, it looks like:

- the `salt://` is mangled to `salt://ad/parameters/dns_domain/dompedago.etb1.lan`
- the destination directory use `ad/parameters/dns:domain/dompedago.etb1.lan`

I think the minion should keep the `salt://` URL and mangle the path used locally.

I'll open an issue in salt.

Thanks a lot @dafyddj.

Author: So, this is not going to work on windows minion until saltstack/salt#58726 is fixed.

I'm not fond of [alternate delimiter](https://docs.saltstack.com/en/latest/topics/targeting/compound.html#alternate-delimiters) since it should be mostly useful when having colon `:` in the value you query by [compound matchers](https://docs.saltstack.com/en/latest/topics/targeting/compound.html).

But I could extend the the `[<TYPE>[:<OPTION>]@]<KEY>` form of the `map.jinja` source definition to `[<TYPE>[:<OPTION>][DELIM]@]<KEY>`

- `Y:G|@dns|domain` will be 
  - `TYPE` is `YAML`
  - `OPTION` is query using `salt["grains.get"]`
  - `DELIM` is `|` which means we need to pass the `delimiter='|'` argument to `salt["config.get"]`
- `Y:@dns:domain` will be
  - `TYPE` is `YAML`
  - `OPTION` is default (query using `salt["config.get"]`)
  - `DELIM` is `:` which means we need to pass the `delimiter=':'` argument to `salt["config.get"]` (which is the default)

It will complicate a little more the code but it looks feasible without too much troubles.

Author: I have an implementation

<details><summary>diff to implement delimiter choice</summary>

``` diff
diff --git a/openssh/map.jinja b/openssh/map.jinja
index 72a33bb..83317e9 100644
--- a/openssh/map.jinja
+++ b/openssh/map.jinja
@@ -138,6 +138,7 @@
     } %}
 
 {#- Process each `map.jinja` source #}
+{#- source format is `[<TYPE>[:<OPTION>][:DELIM]@]<KEY>` #}
 {#- each source has a type: #}
 {#- - `Y` to load values from YAML files (the default when no type is set) #}
 {#- - `C` to lookup values with `config.get` #}
@@ -149,24 +150,47 @@
 {#- - `I` for query with `pillar.get` #}
 {#- The `C`, `G` or `I` types can define the `SUB` option #}
 {#- to merge values in the sub key `mapdata.<key>` instead of directly in `mapdata` #}
+{#- You can choose a different delimiter when doing queries #}
 {%- for map_source in map_sources %}
 {%-   set source_parts = map_source.split('@') %}
 {%-   if source_parts|length == 1 %}
 {#-     By default we load YAML files for config looked up by `config.get` #}
 {%-     set source_type = "Y" %}
 {%-     set query_type = "C" %}
+{%-     set query_delimiter = ":" %}
 {%-     set source_key = map_source %}
-{%-   elif source_parts[0][0] == "Y" %}
-{%-     set source_type = "Y" %}
-{%-     set query_type = source_parts[0].split(':')[1] | default("C") %}
-{%-     set source_key = source_parts[1] %}
-{%-   elif source_parts[0][0] in query_map.keys() %}
-{%-     set source_type = source_parts[0].split(':') | first %}
-{%-     set query_type = source_type %}
-{%-     set is_sub_key = source_parts[0].split(':')[1] | default(False) == "SUB" %}
+{%-   else %}
+{%-     set metadatas = source_parts[0].split(":") %}
 {%-     set source_key = source_parts[1] %}
+{%-     if metadatas|length == 1 %}
+{%-       set source_type = metadatas[0] %}
+{%-       set source_option = "C" %}
+{%-       set query_delimiter = ":" %}
+{%-     elif metadatas|length == 2 %}
+{%-       set source_type = metadatas[0] %}
+{%-       set source_option = metadatas[1] %}
+{%-       set query_delimiter = ":" %}
+{%-     elif metadatas|length == 3 %}
+{%-       set source_type = metadatas[0] %}
+{%-       set source_option = metadatas[1] %}
+{%-       set query_delimiter = metadatas[2] | default(":", boolean=True) %}
+{%-     elif metadatas|length == 4 %}
+{#-       The delimiter is `:` #}
+{%-       set source_type = metadatas[0] %}
+{%-       set source_option = metadatas[1] %}
+{%-       set query_delimiter = ":" %}
+{%-     endif %}
+
+{%-     if source_type == "Y" %}
+{%-       set query_type = source_option | default("C", boolean=True) %}
+{%-     elif source_type in query_map.keys() %}
+{%-       set query_type = source_type %}
+{%-       set is_sub_key = source_option | default(False) == "SUB" %}
+{%-     endif %}
+
 {%-   endif %}
 
+
 {%-   set query_method = query_map[query_type] %}
 
 {%-   if source_type in query_map.keys() %}
@@ -184,11 +208,13 @@
           "map.jinja: retrieve '"
           ~ source_key
           ~ "' with '" ~ query_method ~ "'"
+          ~ ", delimiter='" ~ query_delimiter ~ "'"
           ~ merge_msg
         ) %}
 {%-     set _config_get = salt[query_method](
           source_key,
           default={},
+          delimiter=query_delimiter,
           **merge_opts
         ) %}
 
@@ -229,8 +255,13 @@
           ~ "'" ~ source_key ~ "'"
           ~ " with "
           ~ "'" ~ query_method ~ "'"
+          ~ ", delimiter='" ~ query_delimiter ~ "'"
+        ) %}
+{%-     set map_values = salt[query_method](
+          source_key,
+          delimiter=query_delimiter,
+          default=[]
         ) %}
-{%-     set map_values = salt[query_method](source_key, []) %}
 
 {#-     Mangle `source_key` to use it as literal path #}
 {%-     if map_values | length == 0 %}

```
</details>

So now it's possible to specify an alternate delimiter as a third element of the colon delimited metadatas. I tested several things as you can see in the following logs:

<details><summary>Debug logs of the changes</summary>

```
2020-10-14 09:53:23,281 [salt.loaded.int.module.logmod:42  ][DEBUG   ][14854] map.jinja: load parameters from sources:
- Y@osarch
- Y:@os_family
- Y:G@os
- Y:G:@osfinger
- C::|@ad|lookup
- C@ad
- Y:G:|@dns|domain
- Y:G::@saltversion
- id
2020-10-14 09:53:23,281 [salt.loaded.int.module.logmod:42  ][DEBUG   ][14854] map.jinja: load per formula default values from ad/parameters/defaults.yaml
2020-10-14 09:53:23,286 [salt.fileclient  :1152][DEBUG   ][14854] In saltenv 'base', looking at rel_path 'ad/parameters/defaults.yaml' to resolve 'salt://ad/parameters/defaults.yaml'
2020-10-14 09:53:23,286 [salt.fileclient  :1161][DEBUG   ][14854] In saltenv 'base', ** considering ** path '/var/cache/salt/minion/files/base/ad/parameters/defaults.yaml' to resolve 'salt://ad/parameters/defaults.yaml'
2020-10-14 09:53:23,288 [salt.utils.lazy  :102 ][DEBUG   ][14854] LazyLoaded config.get
2020-10-14 09:53:23,289 [salt.loaded.int.module.logmod:42  ][DEBUG   ][14854] map.jinja: lookup 'osarch' with 'config.get', delimiter=':'
2020-10-14 09:53:23,289 [salt.loaded.int.module.logmod:42  ][DEBUG   ][14854] map.jinja: load parameters from file ad/parameters/osarch/amd64.yaml
2020-10-14 09:53:23,294 [salt.fileclient  :1129][DEBUG   ][14854] Could not find file 'salt://ad/parameters/osarch/amd64.yaml' in saltenv 'base'
2020-10-14 09:53:23,294 [salt.loaded.int.module.logmod:42  ][DEBUG   ][14854] map.jinja: lookup 'os_family' with 'config.get', delimiter=':'
2020-10-14 09:53:23,294 [salt.loaded.int.module.logmod:42  ][DEBUG   ][14854] map.jinja: load parameters from file ad/parameters/os_family/Debian.yaml
2020-10-14 09:53:23,299 [salt.fileclient  :1129][DEBUG   ][14854] Could not find file 'salt://ad/parameters/os_family/Debian.yaml' in saltenv 'base'
2020-10-14 09:53:23,299 [salt.loaded.int.module.logmod:42  ][DEBUG   ][14854] map.jinja: lookup 'os' with 'grains.get', delimiter=':'
2020-10-14 09:53:23,300 [salt.utils.lazy  :102 ][DEBUG   ][14854] LazyLoaded grains.get
2020-10-14 09:53:23,300 [salt.loaded.int.module.logmod:42  ][DEBUG   ][14854] map.jinja: load parameters from file ad/parameters/os/Ubuntu.yaml
2020-10-14 09:53:23,305 [salt.fileclient  :1129][DEBUG   ][14854] Could not find file 'salt://ad/parameters/os/Ubuntu.yaml' in saltenv 'base'
2020-10-14 09:53:23,305 [salt.loaded.int.module.logmod:42  ][DEBUG   ][14854] map.jinja: lookup 'osfinger' with 'grains.get', delimiter=':'
2020-10-14 09:53:23,305 [salt.loaded.int.module.logmod:42  ][DEBUG   ][14854] map.jinja: load parameters from file ad/parameters/osfinger/Ubuntu-18.04.yaml
2020-10-14 09:53:23,310 [salt.fileclient  :1129][DEBUG   ][14854] Could not find file 'salt://ad/parameters/osfinger/Ubuntu-18.04.yaml' in saltenv 'base'
2020-10-14 09:53:23,310 [salt.loaded.int.module.logmod:42  ][DEBUG   ][14854] map.jinja: retrieve 'ad|lookup' with 'config.get', delimiter='|', merge: strategy='None'
2020-10-14 09:53:23,311 [salt.loaded.int.module.logmod:42  ][DEBUG   ][14854] map.jinja: merge 'ad|lookup' retrieved with 'config.get', merge: strategy='smart', lists='False'
2020-10-14 09:53:23,311 [salt.utils.lazy  :102 ][DEBUG   ][14854] LazyLoaded slsutil.merge
2020-10-14 09:53:23,311 [salt.loaded.int.module.logmod:42  ][DEBUG   ][14854] map.jinja: retrieve 'ad' with 'config.get', delimiter=':', merge: strategy='None'
2020-10-14 09:53:23,312 [salt.loaded.int.module.logmod:42  ][DEBUG   ][14854] map.jinja: merge 'ad' retrieved with 'config.get', merge: strategy='smart', lists='False'
2020-10-14 09:53:23,312 [salt.loaded.int.module.logmod:42  ][DEBUG   ][14854] map.jinja: lookup 'dns|domain' with 'grains.get', delimiter='|'
2020-10-14 09:53:23,312 [salt.loaded.int.module.logmod:42  ][DEBUG   ][14854] map.jinja: load parameters from file ad/parameters/dns|domain/example.net.yaml
2020-10-14 09:53:23,317 [salt.fileclient  :1129][DEBUG   ][14854] Could not find file 'salt://ad/parameters/dns|domain/example.net.yaml' in saltenv 'base'
2020-10-14 09:53:23,317 [salt.loaded.int.module.logmod:42  ][DEBUG   ][14854] map.jinja: lookup 'saltversion' with 'grains.get', delimiter=':'
2020-10-14 09:53:23,317 [salt.loaded.int.module.logmod:42  ][DEBUG   ][14854] map.jinja: load parameters from file ad/parameters/saltversion/3001.1.yaml
2020-10-14 09:53:23,322 [salt.fileclient  :1129][DEBUG   ][14854] Could not find file 'salt://ad/parameters/saltversion/3001.1.yaml' in saltenv 'base'
2020-10-14 09:53:23,322 [salt.loaded.int.module.logmod:42  ][DEBUG   ][14854] map.jinja: lookup 'id' with 'config.get', delimiter=':'
2020-10-14 09:53:23,322 [salt.loaded.int.module.logmod:42  ][DEBUG   ][14854] map.jinja: load parameters from file ad/parameters/id/pc-profs-linux.dompedago.etb1.lan.yaml
2020-10-14 09:53:23,326 [salt.fileclient  :1129][DEBUG   ][14854] Could not find file 'salt://ad/parameters/id/pc-profs-linux.dompedago.etb1.lan.yaml' in saltenv 'base'
2020-10-14 09:53:23,327 [salt.loaded.int.module.logmod:42  ][DEBUG   ][14854] map.jinja: save parameters in variable 'mapdata'
```
</details>

Author: On the windows minion I needed to use an exclamation mark `!` because pipe `|` is not working either ;-)

<details><summary>Debug logs of the windows minion</summary>

```
2020-10-14 10:18:33,992 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4536] map.jinja: load parameters from sources:
- Y@osarch
- Y:@os_family
- Y:G@os
- Y:G:@osfinger
- C::|@ad|lookup
- C:@ad
- Y:G:!@dns!domain
- Y:G::@saltversion
- id
2020-10-14 10:18:33,992 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4536] map.jinja: load per formula default values from ad/parameters/defaults.yaml
2020-10-14 10:18:33,992 [salt.fileclient  :1152][DEBUG   ][4536] In saltenv 'base', looking at rel_path 'ad/parameters/defaults.yaml' to resolve 'salt://ad/parameters/defaults.yaml'
2020-10-14 10:18:33,992 [salt.fileclient  :1161][DEBUG   ][4536] In saltenv 'base', ** considering ** path 'c:\salt\var\cache\salt\minion\files\base\ad\parameters\defaults.yaml' to resolve 'salt://ad/parameters/defaults.yaml'
2020-10-14 10:18:34,023 [salt.utils.lazy  :102 ][DEBUG   ][4536] LazyLoaded config.get
2020-10-14 10:18:34,023 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4536] map.jinja: lookup 'osarch' with 'config.get', delimiter=':'
2020-10-14 10:18:34,023 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4536] map.jinja: load parameters from file ad/parameters//osarch.yaml
2020-10-14 10:18:34,023 [salt.fileclient  :1129][DEBUG   ][4536] Could not find file 'salt://ad/parameters//osarch.yaml' in saltenv 'base'
2020-10-14 10:18:34,023 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4536] map.jinja: lookup 'os_family' with 'config.get', delimiter=':'
2020-10-14 10:18:34,023 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4536] map.jinja: load parameters from file ad/parameters/os_family/Windows.yaml
2020-10-14 10:18:34,039 [salt.fileclient  :1129][DEBUG   ][4536] Could not find file 'salt://ad/parameters/os_family/Windows.yaml' in saltenv 'base'
2020-10-14 10:18:34,039 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4536] map.jinja: lookup 'os' with 'grains.get', delimiter=':'
2020-10-14 10:18:34,039 [salt.utils.lazy  :102 ][DEBUG   ][4536] LazyLoaded grains.get
2020-10-14 10:18:34,039 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4536] map.jinja: load parameters from file ad/parameters/os/Windows.yaml
2020-10-14 10:18:34,039 [salt.fileclient  :1129][DEBUG   ][4536] Could not find file 'salt://ad/parameters/os/Windows.yaml' in saltenv 'base'
2020-10-14 10:18:34,039 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4536] map.jinja: lookup 'osfinger' with 'grains.get', delimiter=':'
2020-10-14 10:18:34,039 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4536] map.jinja: load parameters from file ad/parameters/osfinger/Windows-10.yaml
2020-10-14 10:18:34,039 [salt.fileclient  :1129][DEBUG   ][4536] Could not find file 'salt://ad/parameters/osfinger/Windows-10.yaml' in saltenv 'base'
2020-10-14 10:18:34,039 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4536] map.jinja: retrieve 'ad|lookup' with 'config.get', delimiter='|', merge: strategy='None'
2020-10-14 10:18:34,039 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4536] map.jinja: merge 'ad|lookup' retrieved with 'config.get', merge: strategy='smart', lists='False'
2020-10-14 10:18:34,039 [salt.utils.lazy  :102 ][DEBUG   ][4536] LazyLoaded slsutil.merge
2020-10-14 10:18:34,054 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4536] map.jinja: retrieve 'ad' with 'config.get', delimiter=':', merge: strategy='None'
2020-10-14 10:18:34,054 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4536] map.jinja: merge 'ad' retrieved with 'config.get', merge: strategy='smart', lists='False'
2020-10-14 10:18:34,054 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4536] map.jinja: lookup 'dns!domain' with 'grains.get', delimiter='!'
2020-10-14 10:18:34,054 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4536] map.jinja: load parameters from file ad/parameters/dns!domain/example.net.yaml
2020-10-14 10:18:34,054 [salt.fileclient  :1152][DEBUG   ][4536] In saltenv 'base', looking at rel_path 'ad/parameters/dns!domain/example.net.yaml' to resolve 'salt://ad/parameters/dns!domain/example.net.yaml'
2020-10-14 10:18:34,054 [salt.fileclient  :1161][DEBUG   ][4536] In saltenv 'base', ** considering ** path 'c:\salt\var\cache\salt\minion\files\base\ad\parameters\dns!domain\example.net.yaml' to resolve 'salt://ad/parameters/dns!domain/example.net.yaml'
2020-10-14 10:18:34,054 [salt.fileclient  :1179][DEBUG   ][4536] Fetching file from saltenv 'base', ** attempting ** 'salt://ad/parameters/dns!domain/example.net.yaml'
2020-10-14 10:18:34,054 [salt.fileclient  :1207][DEBUG   ][4536] No dest file found
2020-10-14 10:18:34,070 [salt.fileclient  :1295][INFO    ][4536] Fetching file from saltenv 'base', ** done ** 'ad/parameters/dns!domain/example.net.yaml'
2020-10-14 10:18:34,070 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4536] map.jinja: merge parameters from ad/parameters/dns!domain/example.net.yaml
2020-10-14 10:18:34,086 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4536] map.jinja: lookup 'saltversion' with 'grains.get', delimiter=':'
2020-10-14 10:18:34,086 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4536] map.jinja: load parameters from file ad/parameters/saltversion/3001.1.yaml
2020-10-14 10:18:34,086 [salt.fileclient  :1129][DEBUG   ][4536] Could not find file 'salt://ad/parameters/saltversion/3001.1.yaml' in saltenv 'base'
2020-10-14 10:18:34,086 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4536] map.jinja: lookup 'id' with 'config.get', delimiter=':'
2020-10-14 10:18:34,086 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4536] map.jinja: load parameters from file ad/parameters/id/PC-423853.dompedago.etb1.lan.yaml
2020-10-14 10:18:34,086 [salt.fileclient  :1129][DEBUG   ][4536] Could not find file 'salt://ad/parameters/id/PC-423853.dompedago.etb1.lan.yaml' in saltenv 'base'
2020-10-14 10:18:34,086 [salt.loaded.int.module.logmod:42  ][DEBUG   ][4536] map.jinja: save parameters in variable 'mapdata'
```
</details>

.. code-block::

    mkdir -p /srv/salt/TEMPLATE/parameters/dns:domain/
    ln -s dns:domain /srv/salt/TEMPLATE/parameters/domain

We create a configuration for the DNS domain ``example.net`` in ``/srv/salt/TEMPLATE/parameters/dns:domain/example.net.yaml``:

.. code-block:: yaml

    ---
    values:
      config: /etc/template-formula-example-net.conf
    ...

We create another configuration for the DNS domain ``example.com`` in ``/srv/salt/TEMPLATE/parameters/dns:domain/example.com.yaml``:

.. code-block:: yaml

    ---
    values:
      config: /etc/template-formula-{{ grains['os_family'] }}.conf
    ...

> Create per role configuration for ``TEMPLATE`` formula

Reviewer 2: Maybe here `TEMPLATE` and `template` strings should be substituted with `openssh-formula` and context adapted for consistency?

Author: I thought that `map.jinja.rst` would be generic across all formulas. Then, each formula should provide its own [configuration documentation](https://tree.taiga.io/project/myii-saltstack-formulas/epic/72).

``````````````````````````````````````````````````````

Now, we can provides the per role configuration files for the ``TEMPLATE`` formulas under ``/srv/salt/TEMPLATE/parameters/``.

We create the directory for roles:

.. code-block::

    mkdir -p /srv/salt/TEMPLATE/parameters/roles

We will define 2 roles:

- ``template/server``
- ``template/client``

We create a configuration for the role ``template/server`` in ``/srv/salt/TEMPLATE/parameters/roles/template/server.yaml``:

.. code-block:: yaml

    ---
    values:
      config: /etc/template-formula-server.conf
    ...

We create another configuration for the role ``template/client`` in ``/srv/salt/TEMPLATE/parameters/roles/template/client.yaml``:

.. code-block:: yaml

    ---
    values:
      config: /etc/template-formula-client.conf
    ...

Enable roles and the ``dns:domain`` and ``domain`` grains for ``map.jinja``
```````````````````````````````````````````````````````````````````````````

We need to redefine the sources for ``map.jinja`` to load values from our new configuration files, we provide a global configuration for all our minions.

> We create the global parameters file ``/srv/salt/parameters/map_jinja.sls``:

Reviewer: ```suggestion
We create the global parameters file ``/srv/salt/parameters/map_jinja.yaml``:
```
I think?

.. code-block:: yaml

    ---
    values:
      map_jinja:
        sources:
          # default values
          - "Y:G@osarch"
          - "Y:G@os_family"
          - "Y:G@os"
          - "Y:G@osfinger"
          - "C@{{ tplroot ~ ':lookup' }}"
          - "C@{{ tplroot }}"

> 	  # Roles activate/desactivate things

Reviewer: ```suggestion
	  # Roles activate/deactivate things
```

Author: Thanks, I even removed tabulations from the file.

	  # then thing are configured depending of environment
	  # So roles comes before `dns:domain`, `domain` and `id`
          - "Y:C@roles"

          # DNS domain configured (DHCP or resolv.conf)
          - "Y:G@dns:domain"

          # Based on minion ID
          - "Y:G@domain"

          # default values
          - "Y:G@id"
    ...

The syntax is explained later at `Sources of configuration values`_.

Bind roles to minions
`````````````````````

We associate roles `grains`_ to minion using `grains.append`_.

For the servers:

.. code-block::

>     salt 'server-*' grains.append template/server

Reviewer: ```suggestion
    salt 'server-*' grains.append roles template/server
```

For the clients:

.. code-block::

>     salt 'client-*' grains.append template/client

Reviewer: ```suggestion
    salt 'client-*' grains.append roles template/client
```

     - otherwise, load the YAML file named ``salt://{{ tplroot }}/parameters/<KEY>.yaml`` if it exists

   - for ``C``, ``G`` or ``I`` source type, lookup the value of ``salt['<QUERY_METHOD>']('<KEY>')``

#. merge the loaded values with the previous ones using `salt.slsutil.merge`_

There will be no error if a YAML file does not exists, they are all optional.

Configuration values from ``salt['config.get']``
````````````````````````````````````````````````

For sources with of type ``C`` declared in ``map_jinja:sources``, you can configure the ``merge`` option of `salt['config.get']`_ by defining per formula ``strategy`` configuration key (retrieved with ``salt['config.get'](tplroot ~ ':strategy')`` with one of the following values:

- ``recurse`` merge recursively dictionaries. Non dictionary values replace already defined values
- ``overwrite`` new value completely replace old ones

By default, no merging is done, the first value found is returned.

Global view of the order of preferences
```````````````````````````````````````

> To make resumé, here is a complete example of the load order of formula configuration values for an ``AMD64`` ``Ubuntu 18.04`` minion named ``minion1.example.net`` for the ``libvirt`` formula:

Reviewer: ```suggestion
To summarize, here is a complete example of the load order of formula configuration values for an ``AMD64`` ``Ubuntu 18.04`` minion named ``minion1.example.net`` for the ``libvirt`` formula:
```

We create the directory for ``dns:domain`` grain and we add a symlink for the ``domain`` grain which is extracted from the minion ``id``:

.. code-block:: console

    mkdir -p /srv/salt/TEMPLATE/parameters/dns:domain/
    ln -s dns:domain /srv/salt/TEMPLATE/parameters/domain

We create a configuration for the DNS domain ``example.net`` in ``/srv/salt/TEMPLATE/parameters/dns:domain/example.net.yaml``:

.. code-block:: yaml

    ---
    values:
      config: /etc/template-formula-example-net.conf
    ...

We create another configuration for the DNS domain ``example.com`` in ``/srv/salt/TEMPLATE/parameters/dns:domain/example.com.yaml``:

.. code-block:: yaml

    ---
    values:
      config: /etc/template-formula-{{ grains['os_family'] }}.conf
    ...

Create per role configuration for ``TEMPLATE`` formula
``````````````````````````````````````````````````````

Now, we can provides the per role configuration files for the ``TEMPLATE`` formulas under ``/srv/salt/TEMPLATE/parameters/``.

We create the directory for roles:

.. code-block:: console

    mkdir -p /srv/salt/TEMPLATE/parameters/roles

We will define 2 roles:

> - ``template/server``
> - ``template/client``
> 
> We create a configuration for the role ``template/server`` in ``/srv/salt/TEMPLATE/parameters/roles/template/server.yaml``:
> 
> .. code-block:: yaml
> 
>     ---
>     values:
>       config: /etc/template-formula-server.conf
>     ...
> 
> We create another configuration for the role ``template/client`` in ``/srv/salt/TEMPLATE/parameters/roles/template/client.yaml``:
> 
> .. code-block:: yaml
> 
>     ---
>     values:
>       config: /etc/template-formula-client.conf
>     ...

Reviewer: If my understanding is correct, I think this makes more sense for an example:
```suggestion
- ``TEMPLATE/server``
- ``TEMPLATE/client``

We create a configuration for the role ``TEMPLATE/server`` in ``/srv/salt/TEMPLATE/parameters/roles/TEMPLATE/server.yaml``:

.. code-block:: yaml

    ---
    values:
      TEMPLATE:
        config: /etc/template-formula-server.conf
    ...

We create another configuration for the role ``TEMPLATE/client`` in ``/srv/salt/TEMPLATE/parameters/roles/TEMPLATE/client.yaml``:

.. code-block:: yaml

    ---
    values:
      TEMPLATE:
        config: /etc/template-formula-client.conf
    ...
```

Author: You are definitely right.

Author: Note that there is no `tplroot` under `values` in YAML files.

          # Roles activate/deactivate things
          # then thing are configured depending on environment
          # So roles comes before `dns:domain`, `domain` and `id`
          - "Y:C@roles"

          # DNS domain configured (DHCP or resolv.conf)
          - "Y:G:!@dns!domain"

          # Based on minion ID
          - "Y:G@domain"

          # default values
          - "Y:G@id"
    ...

And then, rename the directory:

.. code-block:: console

    mv /srv/salt/TEMPLATE/parameters/dns:domain/  '/srv/salt/TEMPLATE/parameters/dns!domain/'

Format of configuration YAML files
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

When you write a new YAML file, note that it must conform to the following layout:

- a mandatory ``values`` key to store the configuration values
- two optional keys to configure the use of `salt.slsutil.merge`_

  - an optional ``strategy`` key to configure the merging strategy, for example ``strategy: 'recurse'``, the default is ``smart``
  - an optional ``merge_lists`` key to configure if lists should be merged or overridden for the ``recurse`` and ``overwrite`` strategy, for example ``merge_lists: 'true'``

Here is a valid example:

> .. code-block:: yaml
> 
>     ---
>     strategy: 'recurse'
>     merge_lists: 'false'
>     values:
>       pkg:
>         name: 'some-package'
>       config: '/path/to/a/configuration/file'
>     ...
> 
> You can use `Jinja`_ as with any SLS files:
> 
> .. code-block:: yaml
> 
>     ---
>     strategy: 'overwrite'
>     merge_lists: 'true'
>     values:
>       output_dir: /tmp/{{ grains['id'] }}
>     ...

Reviewer: Similarly:
```suggestion
.. code-block:: yaml

    ---
    strategy: 'recurse'
    merge_lists: 'false'
    values:
      TEMPLATE:
        pkg:
          name: 'some-package'
        config: '/path/to/a/configuration/file'
    ...

You can use `Jinja`_ as with any SLS files:

.. code-block:: yaml

    ---
    strategy: 'overwrite'
    merge_lists: 'true'
    values:
      TEMPLATE:
        output_dir: /tmp/{{ grains['id'] }}
    ...
```

Author: There is no `tplroot` under `values` in the YAML files.

Reviewer: I can only go on my experiments but I have to use the `tplroot` in the yaml file to get my parameters to be effective:

https://gist.github.com/dafyddj/35dd3fc7e9bcef47112518c62f582bc7

Author: Erf, in fact it's a specificity of `openssh-formula` which [use subkeys for different components](https://github.com/saltstack-formulas/openssh-formula/pull/191/files#diff-26dfd111adaa5f111e2dd6df3c7c515cabd1eb96c7047695d2b8fc29efd16b6f).

In YAML files, the `values` key match the `mapdata` variable, so the following:

```yaml
values:
  openssh:
    foo: True
```

Is `mapdata.openssh.foo`.

Most formulas are like

```yaml
values:
  parameter: value
```

which became `mapdata.parameter`.

Hope it helps.

Reviewer: Ok, thanks for the explanation. I guess it will all make more sense when we have it working on a few different formulas.

Sources of configuration values
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Configuring ``map.jinja`` sources
`````````````````````````````````

> The ``map.jinja`` file uses several sources where to lookup parameter values. The list of sources can be modified by two files:
> 
> 1. a global ``salt://parameters/map_jinja.yaml``
> 2. a per formula ``salt://{{ tplroot }}/parameters/map_jinja.yaml``.
> 
> Each source definition has the form ``[<TYPE>[:<OPTION>[:<DELIMITER>]]@]<KEY>`` where ``<TYPE>`` can be one of:

Reviewer: ```suggestion
The ``map.jinja`` file uses several sources where to lookup parameter values. The list of sources can be modified by two files:

1. a global ``salt://parameters/map_jinja.yaml``
2. a per formula ``salt://TEMPLATE/parameters/map_jinja.yaml``.

The per formula configuration overrides the global configuration.

Each source definition has the form ``[<TYPE>[:<OPTION>[:<DELIMITER>]]@]<KEY>`` where ``<TYPE>`` can be one of:
```

Author: I put the comment on the `2.` line. Thanks.
